external/avro/src/test/scala/org/apache/spark/sql/avro/AvroFunctionsSuite.scala

```diff
@@ -153,4 +153,79 @@ class AvroFunctionsSuite extends QueryTest with SharedSparkSession {
       assert(df.collect().map(_.get(0)) === Seq(Row("one"), Row("two"), Row("three"), Row("four")))
     }
   }
+
+  test("SPARK-27506: roundtrip in to_avro and from_avro with different compatible schemas") {
+    val df = spark.range(10).select(
+      struct('id.as("col1"), 'id.cast("string").as("col2")).as("struct")
+    )
+    val avroStructDF = df.select(functions.to_avro('struct).as("avro"))
+    val actualAvroSchema = s"""
+                              |{
+                              |  "type": "record",
+                              |  "name": "struct",
+                              |  "fields": [
+                              |    {"name": "col1", "type": "int"},
+                              |    {"name": "col2", "type": "string"}
+                              |  ]
+                              |}
```

Comment:
Indentation seems weird .. I would do:

```scala
val actualAvroSchema =
  s"""
    |
    ...
```

Reply:
+1
@Fokko Could you address this comment?

Reply:
I've fixed this, but something looks off with Github

Reply:
I've also looked through the codebase of Spark, but it is rather incoherent. Please let me know if you want to change it, and I'll update it right away.

Reply:
Yes please update it. I check out the code and in IDE it looks like the way in Github

Reply:
I think it should be as:
```
val msg =
  s"""
     |{
     |
     |
     |}
     |""".stripMargin
```

Reply:
I've updated the PR, feel free to comment.